sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuScalaUDF.scala

```diff
@@ -45,24 +47,50 @@ object GpuScalaUDF {
       repeatingParamCheck = Some(RepeatingParamCheck("param", udfTypeSig, TypeSig.all))),
     (a, conf, p, r) => new ExprMeta[ScalaUDF](a, conf, p, r) {
       override def tagExprForGpu(): Unit = {
-        a.function match {
-          case _: RapidsUDF =>
-          case _ =>
-            val udfName = a.udfName.getOrElse("UDF")
-            val udfClass = a.function.getClass
-            willNotWorkOnGpu(s"$udfName implemented by $udfClass does not provide " +
-                "a GPU implementation")
+        if (getRapidsUDFInstance(a.function).isEmpty) {
+          val udfName = a.udfName.getOrElse("UDF")
+          val udfClass = a.function.getClass
+          willNotWorkOnGpu(s"$udfName implemented by $udfClass does not provide " +
+              "a GPU implementation")
         }
       }
 
       override def convertToGpu(): GpuExpression = {
         GpuScalaUDF(
-          a.function.asInstanceOf[RapidsUDF],
+          getRapidsUDFInstance(a.function).get,
           a.dataType,
           childExprs.map(_.convertToGpu()),
           a.udfName,
           a.nullable,
           a.udfDeterministic)
       }
     })
+
+  /**
+   * Determine if the UDF function implements the [[RapidsUDF]] interface, returning the instance
+   * if it does. The lambda wrapper that Spark applies to Java UDFs will be inspected if necessary
+   * to locate the user's UDF instance.
+   */
+  private def getRapidsUDFInstance(function: AnyRef): Option[RapidsUDF] = {
+    function match {
+      case f: RapidsUDF => Some(f)
+      case f =>
+        try {
+          val clazz = f.getClass
+          val writeReplace = clazz.getDeclaredMethod("writeReplace")
```

Comment:
this might be worth putting more detailed comments on what writeReplace method is and how this is working.

Reply:
I added some comments and also made the check more discriminating by checking the class name to see if it looks like a lambda instance before blindly invoking a `writeReplace` method if we find one.